src/scaled-model.jl

```diff
@@ -0,0 +1,236 @@
+export ScaledModel
+
+struct IpoptScaling{T}
+  max_gradient::T
+end
+
+function _set_constraints_scaling!(cons, Ji, Jj, Jx, max_gradient)
+  # Return a vector storing at index i norm(∇cᵢ, Inf)
+  for (i, j, x) in zip(Ji, Jj, Jx)
+    cons[i] = max(cons[i], abs(x))
+  end
+  # Compute scaling as min(1, max_gradient / norm(∇cᵢ, Inf) )
+  for i in eachindex(cons)
+    cons[i] = min(1.0, max_gradient / cons[i])
+  end
+end
+
+function _set_jacobian_scaling!(Jx, Ji, Jj, cons)
+  k = 0
+  for (i, j) in zip(Ji, Jj)
+    Jx[k += 1] = cons[i]
+  end
+end
+
+function scale_model!(scaling::IpoptScaling{T}, nlp) where T
+  n, m = NLPModels.get_nvar(nlp), NLPModels.get_ncon(nlp)
+  nnzj = NLPModels.get_nnzj(nlp)
+  x0 = NLPModels.get_x0(nlp)
+  g = NLPModels.grad(nlp, x0)
+  scaling_obj = min(one(T), scaling.max_gradient / norm(g, Inf))
+  scaling_cons = similar(x0, m)
+  scaling_jac  = similar(x0, nnzj)
+  fill!(scaling_cons, zero(T))
+  Ji, Jj = NLPModels.jac_structure(nlp)
+  NLPModels.jac_coord!(nlp, x0, scaling_jac)
+  _set_constraints_scaling!(scaling_cons, Ji, Jj, scaling_jac, scaling.max_gradient)
+  _set_jacobian_scaling!(scaling_jac, Ji, Jj, scaling_cons)
+  return (scaling_obj, scaling_cons, scaling_jac)
+end
+
+@doc raw"""
+    ScaledModel
+
+Scale the nonlinear program
+```math
+\begin{aligned}
+       min_x  \quad & f(x)\\
+\mathrm{s.t.} \quad &  c♭ ≤ c(x) ≤ c♯ \\
+                    & x ≥ 0
+\end{aligned}
+```
+as
+```math
+\begin{aligned}
+       min_x  \quad & σf . f(x)\\
+\mathrm{s.t.} \quad &  σc . c♭ ≤ σc . c(x) ≤ σc . c♯ \\
+                    & x ≥ 0
+\end{aligned}
+```
+with ``σf`` a scalar defined as
+```
+σf = min(1, max_gradient / norm(g0, Inf))
+
+```
+and ``σc`` a vector whose size is equal to the number of constraints in the model.
+For ``i=1, ..., m``,
+```
+σc[i] = min(1, max_gradient / norm(J0[i, :], Inf))
+
+```
+
+The vector ``g0 = ∇f(x0)`` and the matrix ``J0 = ∇c(x0)`` are resp.
+the gradient and the Jacobian evaluated at the initial point ``x0``.
+
+"""
+struct ScaledModel{T, S, M} <: NLPModels.AbstractNLPModel{T, S}
+  nlp::M
+  meta::NLPModels.NLPModelMeta{T, S}
+  counters::NLPModels.Counters
+  scaling_obj::T
+  scaling_cons::S # [size m]
+  scaling_jac::S  # [size nnzj]
+  buffer_cons::S  # [size m]
```

Comment:
```suggestion
  scaling_cons::S
  scaling_jac::S
  buffer_cons::S
```
i think you can remove the comments and add
```
@lencheck m scaling_cons buffer_cons
@lencheck nnzj scaling_jac
```
in the constructor.

Reply:
This is just a comment to help the future developers. I don't think there is a need to add `@lencheck` in the constructor as we are initializing the arrays just before.